tests/components/dio_chacon/test_config_flow.py

```diff
@@ -0,0 +1,120 @@
+            DOMAIN,
+            context={"source": config_entries.SOURCE_USER},
+            data={
+                CONF_USERNAME: "dummylogin",
+                CONF_PASSWORD: "dummypass",
+                CONF_UNIQUE_ID: "dummy-user-id",
+            },
+        )
+
+        _LOGGER.debug("Test result after init : %s", result)
+
+    assert len(mock_dio_chacon_client.mock_calls) == 1
+
+    assert result["type"] == FlowResultType.CREATE_ENTRY
+    assert result["title"] == "Dio Chacon dummylogin"
+    assert result["data"] == {
+        CONF_USERNAME: "dummylogin",
+        CONF_PASSWORD: "dummypass",
+        CONF_UNIQUE_ID: "dummy-user-id",
+    }
+
+    assert len(mock_setup_entry.mock_calls) == 1
+
+
+async def test_form_cannot_connect(
+    hass: HomeAssistant, mock_setup_entry: AsyncMock
+) -> None:
+    """Test we handle cannot connect error."""
+    with patch(
+        "dio_chacon_wifi_api.DIOChaconAPIClient.get_user_id",
+        side_effect=DIOChaconAPIError,
+    ) as mock_dio_chacon_client:
+        result = await hass.config_entries.flow.async_init(
+            DOMAIN,
+            context={"source": config_entries.SOURCE_USER},
+            data={
+                CONF_USERNAME: "nada",
+                CONF_PASSWORD: "nadap",
+            },
+        )
+
+    assert result["type"] == FlowResultType.FORM
+    assert result["errors"] == {"base": "cannot_connect"}
+    assert len(mock_dio_chacon_client.mock_calls) == 1
+
+
+async def test_form_invalid_auth(
+    hass: HomeAssistant, mock_setup_entry: AsyncMock
+) -> None:
+    """Test we handle cannot connect error."""
+    with patch(
+        "dio_chacon_wifi_api.DIOChaconAPIClient.get_user_id",
+        side_effect=DIOChaconInvalidAuthError,
+    ) as mock_dio_chacon_client:
+        result = await hass.config_entries.flow.async_init(
+            DOMAIN,
+            context={"source": config_entries.SOURCE_USER},
+            data={
+                CONF_USERNAME: "nada",
+                CONF_PASSWORD: "nadap",
+            },
+        )
+
+    assert result["type"] == FlowResultType.FORM
+    assert result["errors"] == {"base": "invalid_auth"}
+    assert len(mock_dio_chacon_client.mock_calls) == 1
+
+
+async def test_other_error(hass: HomeAssistant, mock_setup_entry: AsyncMock) -> None:
+    """Test we handle any error."""
+    with patch(
+        "dio_chacon_wifi_api.DIOChaconAPIClient.get_user_id",
+        side_effect=Exception("Bad request Boy :) --"),
+    ) as mock_dio_chacon_client:
+        result = await hass.config_entries.flow.async_init(
+            DOMAIN,
+            context={"source": config_entries.SOURCE_USER},
+            data={
+                CONF_USERNAME: "nada",
+                CONF_PASSWORD: "nadap",
+            },
+        )
+
+    assert result["type"] == FlowResultType.FORM
+    assert result["errors"] == {"base": "unknown"}
+    assert len(mock_dio_chacon_client.mock_calls) == 1
```

Comment:
All these tests can be merged by using `parametrize`. You will then make the exception raised and the error message a parameter and create a generic test.

Also please make sure these tests end in a CREATE_ENTRY. This way we are also unit testing a config flow is recoverable after an error.

Reply:
Done